### PR TITLE
Test all operators with assert_valid, or fail with a reason

### PR DESCRIPTION
**Pull request by timmysilv**
**Context:**
Continuation of the work I started in #4922 (this is part 1 of the future work listed in that PR)

**Description of the Change:**
1. Implement `_flatten` for `BlockEncode` and `SpecialUnitary` to pytreeify those types
2. Organize `tests/ops/functions/conftest.py` to better parametrize tests. The parametrization fixtures are easier to make using the following:
  - `_INSTANCES_TO_TEST` is a list of operator instances that couldn't be auto-generated, but do pass validation
  - `_INSTANCES_TO_FAIL` is a list of tuples (`(Operator, Type[Exception])`) where the first element is an operator instance that will fail validation, and the second element is the exception type raised during validation. Ideally, it also comes with a comment explaining what's failing/why it should fail.

All test collection logic is now in `conftest.py`, making the tests much smaller and simpler. The only "extra" logic in the tests is the xfail I added for all qutrit ops before auto-generation. Now there are 3 tests:
1. `test_generated_list_of_ops`, which auto-generates instances of every Operator type and asserts its validity
2. `test_explicit_list_of_ops`, which tests the validity of manual operator instances
3. `test_explicit_list_of_failing_ops`, which ensures that invalid operator instances fail to validate as expected

**Benefits:**
Every single pennylane operator (except for templates) is (in)validated in a single test file, and most of it is done automatically.

**Possible Drawbacks:**
Yet another test file that needs maintenance.

### CODE REVIEW

File: doc/releases/changelog-dev.md
Comment (timmysilv) [inline doc/releases/changelog-dev.md:10]:
```suggestion
  [(#4978)](https://github.com/PennyLaneAI/pennylane/pull/4978)
```

File: tests/ops/functions/conftest.py
Comment (albi3ro) [inline tests/ops/functions/conftest.py:61]: What would you think of expanding the definition of `TensorLike` to include `scipy.sparse.sparray` and `scipy.sparse.spmatrix`? Seems to work out for me locally. 

We then run into the issue that `allclose` doesn't work for sparse mats, but we can probably fix that too.

File: tests/ops/functions/conftest.py
Comment (albi3ro) [inline tests/ops/functions/conftest.py:64]: eek... how is that not causing problems?

File: tests/ops/functions/conftest.py
Comment (albi3ro) [inline tests/ops/functions/conftest.py:96]: pickling error now resolved on release candidate.

File: tests/ops/functions/conftest.py
Comment (albi3ro) [inline tests/ops/functions/conftest.py:100]: We could skip the wire mapping test if the operator has more than 26 wires?

File: tests/ops/functions/conftest.py
Comment (albi3ro) [inline tests/ops/functions/conftest.py:92]:
```suggestion
        qml.pulse.ParametrizedEvolution(qml.PauliX(0) + sum * qml.PauliZ(0)),
        ValueError,  # binding parameters fail, and more
```

File: tests/ops/functions/conftest.py
Comment (timmysilv) [reply tests/ops/functions/conftest.py:61]: I'm ok with it, but it will probably take more work as you're suggesting with the benefit pretty much just being us validating `SparseHamiltonian`

**Review by albi3ro — APPROVED** [review]
I obviously like having a list of all the things we need to fix. Though I am distracted by wanting to fix all those issues immediately, the point of isn't to fix the issues, but just to keep a list of the things we need to fix.

**Review by mudit2812 — APPROVED** [review]
Looks good, but lets wait on merging this until after the RC-master merge and fixing the consequent `IsingXX` `PickleError` (or lack thereof).

File: tests/ops/functions/conftest.py
Comment (timmysilv) [reply tests/ops/functions/conftest.py]: done

**Pull request merged**
